src/libpmemobj/backend_persistent.c

```diff
@@ -0,0 +1,464 @@
+/*
+ * Copyright (c) 2015, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of Intel Corporation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * backend_persistent.c -- implementation of persistent backend
+ */
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <libpmem.h>
+#include <string.h>
+#include "bucket.h"
+#include "arena.h"
+#include "backend.h"
+#include "pool.h"
+#include "util.h"
+#include "out.h"
+#include "backend_persistent.h"
+
+static struct bucket_backend_operations persistent_bucket_ops = {
+
+};
+
+static struct arena_backend_operations persistent_arena_ops = {
+	.set_alloc_ptr = persistent_set_alloc_ptr
+};
+
+static struct pool_backend_operations persistent_pool_ops = {
+
+};
+
+/*
+ * verify_header -- (internal) check is the header is consistent
+ */
+static bool
+verify_header(struct backend_pool_header *h)
+{
+	if (util_checksum(h, sizeof (*h), &h->checksum, 0) != 1) {
+		return false;
+	}
+
+	if (memcmp(h->signature, POOL_SIGNATURE, POOL_SIGNATURE_LEN) != 0) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * copy_header -- (internal) create a copy of a header
+ */
+static void
+copy_header(struct backend_persistent *b, struct backend_pool_header *left,
+	struct backend_pool_header *right)
+{
+	memcpy(left, right, sizeof (*left));
```

Comment:
Why not use pmem_memcpy() here, and not call b->persist? Same for other cases where this is done.